app/src/main/java/org/p2p/wallet/user/repository/UserInMemoryRepository.kt

```diff
@@ -3,60 +3,93 @@ package org.p2p.wallet.user.repository
 import org.p2p.wallet.home.model.TokenPrice
+import org.p2p.wallet.receive.list.TokenListData
 import org.p2p.wallet.user.model.TokenData
 import timber.log.Timber
 
+private const val DEFAULT_TOKEN_KEY = "DEFAULT_TOKEN_KEY"
+
 class UserInMemoryRepository : UserLocalRepository {
     private val popularItems = arrayOf("SOL", "USDC", "BTC", "USDT", "ETH")
     private val pricesFlow = MutableStateFlow<List<TokenPrice>>(emptyList())
     private val tokensFlow = MutableStateFlow<List<TokenData>>(emptyList())
 
+    private val tokensSearchResultFlow = MutableStateFlow<TokenListData>(TokenListData())
+    private val searchTextByTokens: MutableMap<String, List<TokenData>> = mutableMapOf()
+
     override fun setTokenPrices(prices: List<TokenPrice>) {
         pricesFlow.value = prices
     }
 
     override fun getPriceByToken(symbol: String): TokenPrice? =
         pricesFlow.value.firstOrNull { it.tokenSymbol == symbol }
 
-    private val decimalsFlow = MutableStateFlow<List<TokenData>>(emptyList())
+    private val allTokensFlow = MutableStateFlow<List<TokenData>>(emptyList())
 
     override fun setTokenData(data: List<TokenData>) {
-        decimalsFlow.value = data.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })
+        allTokensFlow.value = data.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.name })
     }
 
     override fun fetchTokens(searchText: String, count: Int, refresh: Boolean) {
         if (refresh) {
-            tokensFlow.value = emptyList()
+            searchTextByTokens.clear()
+            tokensSearchResultFlow.value = TokenListData()
         }
-        if (tokensFlow.value.size >= decimalsFlow.value.size) {
+        val allInMemoryTokens = allTokensFlow.value
+        val currentSearchTokenResultSize = tokensSearchResultFlow.value.size
+        if (currentSearchTokenResultSize >= allInMemoryTokens.size) {
             return
         }
-        val items = decimalsFlow.value
-        val offset = tokensFlow.value.size
 
-        if (searchText.isNotEmpty()) {
-            val result = items.filter {
-                searchText == it.symbol ||
-                    it.name.startsWith(searchText, ignoreCase = true)
+        when {
+            searchText.isEmpty() -> {
+                val defaultTokens = searchTextByTokens.getOrDefault(DEFAULT_TOKEN_KEY, emptyList())
+                val newSize = (defaultTokens.size - getPopularDecimals().size) + count
+                val needToLoadMore = newSize > defaultTokens.size
+                if (needToLoadMore) {
+                    searchTextByTokens[DEFAULT_TOKEN_KEY] = getPopularDecimals() + allInMemoryTokens.take(newSize)
+                }
+                setResult(DEFAULT_TOKEN_KEY)
+            }
+            searchText in searchTextByTokens -> {
+                val tokensBySearchText = searchTextByTokens.getOrDefault(searchText, emptyList())
+                val newSize = tokensBySearchText.size + count
```

Comment:
maybe `count` is not a good choice of words here
what count really means? 
it's like new search list size, right? maybe we can think in this direction and name it better